DOC: add documentation on why astropy has tables

taldcroft opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
https://docs.astropy.org/en/latest/development/quickstart.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/git_edit_workflow_examples.html . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request is to address a common question of why astropy maintains its own Table class instead of using pandas DataFrame.

It also tightens up the opening description and feature list.

This also substantially updates the page on `Interfacing with the Pandas Package`. Some of that was out of date or just wrong, so I fixed things and added a detailed example highlighting known issues.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #17146

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

mhvk reviewed (commented):
I like the text very much. My main question is whether this is not better as a separate page, as in the main text it seems a bit distracting. A separate page would also more easily allow (others) to add more detailed information about transformations between `QTable` and `pandas`. Maybe a new `interoperability.rst`?

taldcroft commented:
@mhvk - good idea with the new page. I also took the opportunity to make this more positive at the top level. More of a "Why astropy table" rather than "Why not pandas", along with the relationship between the two. Also added a mention of polars at the end.

taldcroft commented:
@mhvk - I also substantially updated the page on `Interfacing with the Pandas Package`. Some of that was out of date or just wrong, so I fixed things and added a detailed example highlighting known issues.

mhvk commented on docs/table/pandas.rst:
```diff
@@ -53,17 +53,124 @@ It is also possible to create a table from a :class:`~pandas.DataFrame`::
+  <https://pandas.pydata.org/docs/user_guide/timeseries.html>`_, but this object has
+  only 64-bit precision and does not support leap seconds or time scales.
+
+These issues a highlighted below in a more complex example with a table that includes
```
`issues a` -> `issues are`

taldcroft replied:
Fixed

mhvk reviewed (approved):
This looks great! I found a small typo, but otherwise I think we should just get it in.

taldcroft commented:
Thanks for the quick review @mhvk. Auto-merge is enabled.

taldcroft commented:
@mhvk - this is failing tests because of the new functions `matvec` and `vecmat` in numpy 2.2.0 that was just released.

neutrinoceros commented:
Patched in #17523

bsipocz commented on docs/table/table_and_dataframes.rst:
```diff
@@ -0,0 +1,95 @@
+write it back out, the data will be effectively identical. This is made possible by
+robust support for table and column metadata which allows storing and propagating common
+column information such as the unit, description, and format. For VOTable data, more
+esoteric information like the UCD is maintained.
```
can we use another, more neutral adjective here? Or even just drop the adjective
```suggestion information like the UCD is maintained.
```

taldcroft replied:
Fixed.

bsipocz commented on docs/table/table_and_dataframes.rst:
```diff
@@ -0,0 +1,95 @@
+column information such as the unit, description, and format. For VOTable data, more
+esoteric information like the UCD is maintained.
+
+FITS and VOTable are two important file formats in astronomy, but the idea of lossless
```
This is absolute nitpicking, but my understanding is that VOTable is a format standard and not a file format standard, and serialization in a format other than XML is theoretically possible. 

Less nitpicky, is that I don't really understand the point this paragraph tries to make in regard of astroquery

taldcroft replied:
Fixed both, hopefully.

bsipocz commented on docs/table/table_and_dataframes.rst:
```diff
@@ -0,0 +1,95 @@
+- :ref:`Time formats <time-format>` used in astronomy such as the FITS time format are
+  not supported.
+
+Pandas does not support coordinate columns.
```
this is significant, maybe you want to divide this section into two, one for time and one for coordinates?

bsipocz reviewed (commented):
I have some comments, mostly for clarification.

bsipocz commented:
btw, just a small suggestion: if review is requested over the weekend it would be a courtesy of not setting up auto-merge before that very same weekend ended.

pllim commented:
I disabled auto merge for now since there are new comments. FYI.

taldcroft reviewed (commented):
@bsipocz - thanks for the comments, I think I've addressed them. Sorry about the auto-merge, I got a bit excited to get this in.

bsipocz commented on docs/table/table_and_dataframes.rst:
```diff
@@ -39,15 +39,17 @@ VOTable data. This means that when you read a FITS or VOTable file into a table
-<https://astroquery.readthedocs.io/en/latest/>`_.
+FITS and VOTable are two important data formats in astronomy, but the idea of lossless
+representation of physical data is critical in other contexts as well. For
+example, accessing metadata like units is a key component of using the
+`astroquery <https://astroquery.readthedocs.io/en/latest/>`_ service for astronomical
+data analysis.
```
I still don't get the astroquery reference, and why this is a key component for astroquery. We could return whatever comes back from the server, typically a VOTable or csv or whatever text file object. It's just a compatibility thing that we went a step further and parse the results into an astropy Table. 


(and returning pandas definitely brought up by a couple of modules but we pushed them to do astropy Tables instead to be more consistent with the older modules).

taldcroft replied:
IMO a key feature of astroquery is exactly the hard work of parsing heterogenous results from data providers into a uniform and predictable data structure (QTable) that is easy for users to work with. Many of those datasets include physical quantities with units, column descriptions, or other metadata, so if astroquery returned a DataFrame then that information would be lost.

That said, if you really don't want this shout-out to astroquery then I can replace it with a different example.

mhvk replied:
Seconded - for me part of the beauty of astroquery is exactly that it produces a table with usable columns!

bsipocz replied:
Yes, I think if you can go with a different example it would be OK as I'm just not comfortable with the 
` is a key component of using the astroquery service for astronomical data analysis` summary, and nitpicking the sentence away is not very constructive.

pllim commented on docs/table/table_and_dataframes.rst:
```diff
@@ -0,0 +1,97 @@
+- :ref:`Time formats <time-format>` used in astronomy such as the FITS time format are
+  not supported.
+
+*Pandas does not support coordinate columns.*
```
```suggestion
*Pandas does not support sky coordinate columns.*
```

pllim commented on docs/table/table_and_dataframes.rst:
```diff
@@ -0,0 +1,97 @@
+We recognize that Pandas is a popular library and that there are many users who are
+familiar with it. For this reason, we have made it easy to convert between
+`astropy.table` and `~pandas.DataFrame`, as documented in :ref:`pandas`. This allows
+users to take advantage of the features of both packages as needed.
```
```suggestion
users to take advantage of the features of both packages as needed,
within the limitations stated above.
```

pllim commented on docs/table/pandas.rst:
```diff
@@ -53,17 +53,124 @@ It is also possible to create a table from a :class:`~pandas.DataFrame`::
+  masked column to a pandas DataFrame, the original values are lost.
+
+* Tables with :ref:`mixin_columns` such as `~astropy.time.Time`,
+  `~astropy.coordinates.SkyCoord`, and `~astropy.units.Quantity` can be converted, but
```
These two forms are interchangable right? I see one form here and one form there; should we just pick one and stick with it?
```suggestion
  `~astropy.coordinates.SkyCoord`, and |Quantity| can be converted, but
```

taldcroft commented on docs/table/table_and_dataframes.rst:
```diff
@@ -0,0 +1,97 @@
+Lossless representation of FITS and VOTable data via metadata
+-------------------------------------------------------------
+
+The `astropy.table` package is designed to provide lossless representation of FITS and
```
Hence the "is designed" qualifier. We could go to something one level lower like "strives to".

pllim replied:
FWIW I would prefer "strives to" because I cannot tell from the text if this "design" is fulfilled or not.

taldcroft replied:
Can you make a suggestion. I'm committing all the rest of yours.

pllim replied:
Suggestion provided in https://github.com/astropy/astropy/pull/17520/files#diff-4bd7af647ce2840ff6686f1fbccf9e767f22776f4e647cef7765df8066f8cce6 . Thanks!

pllim commented on docs/table/index.rst:
```diff
@@ -358,6 +356,14 @@ Mixin Columns
 
    mixin_columns.rst
 
+Astropy Table and DataFrames
```
Should we at least move it to just after `pandas.rst`? Seems jarring to have two topics touching pandas (which is an optional dependency) separated by something else in between.

pllim commented on docs/table/table_and_dataframes.rst:
```diff
@@ -0,0 +1,97 @@
+Lossless representation of FITS and VOTable data via metadata
+-------------------------------------------------------------
+
+The `astropy.table` package is designed to provide lossless representation of FITS and
```
```suggestion
The `astropy.table` package strives to provide lossless representation of FITS and
```

taldcroft commented:
OK all the @pllim comments are addressed, I think.

mhvk commented on docs/table/index.rst:
```diff
@@ -7,24 +7,22 @@ Data Tables (`astropy.table`)
+* :ref:`Read and write <read_write_tables>` to files via the :ref:`Unified File Read/Write Interface <table_io>`.
+* Convert to and from `pandas.DataFrame`.
+
+The :ref:`astropy-table-and-dataframes` page provides the rationale for maintaining
```
I'd move this as part of the item just above -- maybe 
```
* Convert to and from `pandas.DataFrame` (see :ref:`pandas`
 as well as :ref:`astropy-table-and-dataframes` for the rationale
 for a dedicated `astropy.table` package instead of relying on `pandas`).
```
(not sure I got the references quite right - want to have a direct link to `pandas.rst` as well...)

p.s. Fine to ignore!

taldcroft replied:
Looking at the rendered HTML again, I saw that every list item was succinct and fit on one line, with the exception of the one about mixins. So instead of making the pandas line longer, I shortened the mixin line and now it looks very clean.

This of course leaves the part about astropy-table-and-dataframes as a distinct coda to the introduction, but this comes up so often that it seems OK.

pllim reviewed (approved):
Thanks!

mhvk reviewed (approved):
Reread, and like it very much still! One suggestion for the stop, to make the `pandas` reference just one item.
